Adding do your own map

Frenchris opened the pull request.

kigiri commented on exercise/do-your-own-map.js:
```diff
@@ -0,0 +1,73 @@
+
+### Instructions
+
+Create a `MapFunc` function that takes an array as first argument, a function as second,
```
no caps in function names, also avoid saying it's a function, so the name would be `map`

By conventions, capitalized functions are constructors in javascript, like the actual `Map` which is a data-struct constructor, let's just call this one `map`, we could add `flatMap` too just for fun.

Frenchris replied:
noted for the no cap comment.
Added flatMap.

kigiri commented on exercise/do-your-own-map.js:
```diff
@@ -0,0 +1,73 @@
+
+const randomInt1 = getRandomInt(100)
+const randomInt2 = getRandomInt(100)
+const randomInt3 = getRandomInt(100)
```
Not sure why we want / need random values here.

If we do need it, please move them in the test context, see my comment here https://github.com/nan-academy/js-training/pull/35#discussion_r413778564

Frenchris replied:
removed the random values for this exercise.

kigiri commented on exercise/do-your-own-map.js:
```diff
@@ -0,0 +1,73 @@
+  )
+)
+
+t(({ code }) => !code.includes('map'))
```
not a solid way to prevent to use map:

```js
// this works perfectly, it should not
const MapFunc = (f => f.call.bind(f))([]['m'+'ap'])
```

it's best to use:
```js
Array.prototype.map = undefined
```
to completly prevent them from using it.

Frenchris replied:
Noted and being applied to other Prs

kigiri commented on exercise/do-your-own-map.js:
```diff
@@ -0,0 +1,73 @@
+
+t(({ eq }) =>
+  eq(
+    MapFunc(arr1, (el) => doAll(el)),
```
test should ensure that we pass the 2nd (index) and 3rd (array) arguments too, like a real map.

it's good to know about those.

Frenchris replied:
Transformed the subject to get the student to use index and array as arguments too.

Frenchris committed:
Module updated for "do your own map" exercise to work

Frenchris committed:
Updated subject for "do your own map"

Frenchris committed:
Previous version of solutions along with working one (frenchris)

kigiri commented on exercise/do-your-own-map.js:
```diff
@@ -0,0 +1,113 @@
+    '20 is at index: 3 of array 5,4,-3,20,17,-33,-4,18',
+    '17 is at index: 4 of array 5,4,-3,20,17,-33,-4,18',
+    '-33 is at index: 5 of array 5,4,-3,20,17,-33,-4,18',
+    '-4 is at index: 6 of array 5,4,-3,20,17,-33,-4,18',
```
a better way to tests if they pass the array, that would also demonstrate the use of `index`:
Do something on every items but the last, example
```js
['a', 'b', 'c']
  .map((item, index, arr) => index === arr.length - 1 ? item : `${item}, `)
  .join('')
// output -> 'a, b, c'
```

It's best to always try to demonstrate actual usecases as much as possible

Frenchris replied:
I updated my function with the arr and added another one inspired from your example.
I added 3 extra arrays and tests.

kigiri commented on exercise/do-your-own-map.js:
```diff
@@ -0,0 +1,113 @@
+
+let arr1 = [10, -10, 20, -95, 86, 102, 35, 89, 110]
+let arr2 = [[10], -10, 20, -95, 86, [102], [35, 89], 110]
+let arr3 = [[5], [4], [-3], [20], [17], [-33], [-4], [18]]
```
why `let` ?
also, those would be better in the `setup`

Frenchris replied:
force of habbit, did not think that one through. I swiched them to `const`

kigiri commented on exercise/do-your-own-map.js:
```diff
@@ -0,0 +1,113 @@
+const sub3 = (el) => el - 3
+
+const doAll = (el) => sub3(mult2(add1(el)))
+const posValsIndex = (el, i) => (el >= 0 ? `${i}: ${el}` : null)
```
please refrain for using `null` as much as possible, use `undefined`.

kigiri replied:
read more about why: https://github.com/sindresorhus/meta/issues/7 (more in the comments)

Frenchris replied:
ok, :) I will refrain from using `null` in my JS code. Interesting thread, did not expect to see so many opinions on the subject  :O  
All `null` have been replaced by `undefined`

kigiri commented on tests/do-your-own-map/pass_frenchris.js:
```diff
@@ -0,0 +1,7 @@
+  return result
+}
+
+const flatMap = (el, fn) => map(el, fn).flat()
```
Should we authorize `flat` ?
Don't that make the exercise too easy ?

Frenchris replied:
Fair point, I restricted [].flat() and updated my solution accordingly

Frenchris committed:
Update of "do your own map" subject
1)Moved the data arrays in the setup module.
2)Changed all the nulls as undefined.
3)Removed the use of flat to increase the difficulty of flatMap.
4)Modified the function using the arr passed as argument to use
the arr.length.
5)Added another function using the arr.length several times.
6)Added three extra tests with the last function created.

Frenchris committed:
Updated solution to reflect the restriction of the [].flat() function

kigiri committed:
call-me-maybe/map: rename and add failing test for for..in
